usecases/backup/coordinator.go

```diff
@@ -0,0 +1,374 @@
+	}()
+
+	return nil
+}
+
+// canCommit asks candidates if they agree to participate in DPRO
+// It returns and error if any candidates refuses to participate
+func (c *coordinator) canCommit(ctx context.Context, method Op) (map[string]struct{}, error) {
+	ctx, cancel := context.WithTimeout(ctx, c.timeoutCanCommit)
+	defer cancel()
+	type pair struct {
+		n string
+		r *Request
+	}
+
+	id, backend := c.descriptor.ID, c.descriptor.Backend
+	groups := c.descriptor.Nodes
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(_MaxNumberConns)
+	reqChan := make(chan pair)
+	g.Go(func() error {
+		defer close(reqChan)
+		for node, gr := range groups {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
+			reqChan <- pair{node, &Request{
+				Method:   method,
+				ID:       id,
+				Backend:  backend,
+				Classes:  gr.Classes,
+				Duration: _BookingPeriod,
+			}}
+		}
+		return nil
+	})
+
+	mutex := sync.RWMutex{}
+	nodes := make(map[string]struct{}, len(groups))
+	for pair := range reqChan {
+		pair := pair
+		g.Go(func() error {
+			resp, err := c.client.CanCommit(ctx, pair.n, pair.r)
+			if err == nil && resp.Timeout == 0 {
+				err = errCannotCommit
+			}
+			if err != nil {
+				return fmt.Errorf("node %q: %w", pair.n, err)
+			}
+			mutex.Lock()
+			nodes[pair.n] = struct{}{}
+			mutex.Unlock()
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		c.abortAll(ctx, id, method, nodes)
+		return nil, err
+	}
+	return nodes, nil
+}
+
+// commit tells each participant to commit its backup operation
+// It stores the final result in the provided backend
+func (c *coordinator) commit(ctx context.Context, req *StatusRequest, nodes map[string]struct{}) {
+	c.commitAll(ctx, req, nodes)
+
+	for len(nodes) > 0 {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(c.timeoutNextRound):
+			c.queryAll(ctx, req, nodes)
+		}
+	}
+	c.descriptor.CompletedAt = time.Now().UTC()
+	status := backup.Success
+	reason := ""
+	groups := c.descriptor.Nodes
+	for node, p := range c.Participants {
+		st := groups[node]
+		st.Status, st.Error = p.Status, p.Reason
+		if p.Status != backup.Success {
+			status = backup.Failed
+			reason = p.Reason
+		}
+		groups[node] = st
+	}
+	c.descriptor.Status = status
+	c.descriptor.Error = reason
+}
+
+// queryAll queries all participant and store their statuses internally
+//
+// It returns the number of remaining nodes to query in the next round
+func (c *coordinator) queryAll(ctx context.Context, req *StatusRequest, nodes map[string]struct{}) int {
+	ctx, cancel := context.WithTimeout(ctx, c.timeoutQueryStatus)
+	defer cancel()
+
+	rs := make([]partialStatus, len(nodes))
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(_MaxNumberConns)
+	i := 0
+	for node := range nodes {
+		j := i
+		rs[j].node = node
+		g.Go(func() error {
+			rs[j].StatusResponse, rs[j].err = c.client.Status(ctx, rs[j].node, req)
+			return nil
+		})
+		i++
+	}
+	g.Wait()
```

Comment:
No error check here

Reply:
No need to check error here because we capture errors inside the go routines (see line 274)